### ISSUE TITLE
Evidence of NAND being used as temp before uploading. It still hurts NAND endurance unless we find the temp location.

### ISSUE THREAD

**Issue by vegedb** [master-hax/pixel-backup-gang#30]
### Tested a 2GB file and upload through file browser.
![image](https://github.com/user-attachments/assets/a97b794f-1054-4305-b35b-8d216744c6a8)

### After 5 minutes of waiting, 
![image](https://github.com/user-attachments/assets/b4b56a58-7522-40d3-83ce-71f58aca47a7)

## Info
The sequence,

1. File browser > share > upload
2. [Network activity = 0MB/s] 5 Minutes from SSD>Phone eMMC NAND temp folder
3. [Network activity = 25MB/s] After 5 minutes, upload starts. Phone eMMC NAND temp folder > Cloud

Hope you could find what's on sda or sda35 that's using as a temp folder. G
Got the time to investigate nand usage and its 100% being used as a "middleman".
I have zero knowledge on where it might be, hope you can help us hunt it down so we can mount an ssd on it.

**Comment by master-hax** [comment]
it is most likely in `/data/data/com.google.android.apps.photos` directory

**Comment by vegedb** [comment]
> it is most likely in `/data/data/com.google.android.apps.photos` directory

Since I'm using 32gb pixel, it has only 20gb free space.
Uploading 20GB file fails, and it shows "could not find media" at the bottom of the screen.
Hope we have another way to upload bypassing the built in storage.

**Comment by rwahyudi** [comment]
@vegedb Can you see if you can reproduce the same behaviour if you bypass the file browser? ie connect the SSD drive to a PC,  copy files directly to the SSD disk, and then mount the SSD back into the pixel.

**Comment by master-hax** [comment]
i did some investigation with `inotifywait` - i think all the media gets written to  `/data/data/com.google.android.apps.photos/databases/gphotos0.db` in chunks before being uploaded

**Comment by nhan000** [comment]
> Since I'm using 32gb pixel, it has only 20gb free space.
Uploading 20GB file fails, and it shows "could not find media" at the bottom of the screen AFTER 10 minutes of WRITES. I assume it fills up the temp folder and bust error once it reaches a limit.

Google Photos per the support website has a [10 GB file size limit for videos](https://support.google.com/photos/answer/6174018).

![Image](https://github.com/user-attachments/assets/12d259df-2537-41cf-bd7b-f97ee5e72e6a)

However, [here](https://developers.google.com/photos/library/guides/upload-media) 20 GB file size can be accepted? It's an API library so might be different.

![Image](https://github.com/user-attachments/assets/2645dcbb-419c-4c38-a35c-7c80e515cdda)

I believe you can get the Pixel to have more than 20 GB free space, provided that you uninstalled/disabled/clear storage and cache of apps quite aggressively.

![Image](https://github.com/user-attachments/assets/f0c67bdd-0a34-4956-8784-67f5bde5dcf4)

So @vegedb, see if it's possible for you to clear up more space and then try re-upload the 20 GB video, or try it with a 15 GB video to see if it's the limit Google Photos impose per the support website.

Also, I'm a noob so I don't know much about the technical details you guys are discussing. Hope I contributed something to the discussion.

> i did some investigation with `inotifywait` - i think all the media gets written to `/data/data/com.google.android.apps.photos/databases/gphotos0.db` in chunks before being uploaded

I only now noticed that it's pretty much conclusive that the files are written into phone's memory before being uploaded?

**Comment by vegedb** [comment]
> So [@vegedb](https://github.com/vegedb) can try to clear up more space and try re-upload the 20 GB video, or try it with a 15 GB video to see if it's the limit Google Photos impose per the support website.
> 
> Also, I'm a noob so I don't know much about the technical details you guys are discussing. Hope I contributed something to the discussion.
> 
> > i did some investigation with `inotifywait` - i think all the media gets written to `/data/data/com.google.android.apps.photos/databases/gphotos0.db` in chunks before being uploaded
> 
> I only now noticed that it's pretty much conclusive that the files are written into phone's memory before being uploaded?

My biggest file is a 17gb 1080p video if it helps. I also have a bunch of videos above 10GB .
So the limit is 20GB? idk maybe someone could test above that.

Yup its 100% using nand first. Until there's new developments, the only thing we can do now is just buy another pixel when it breaks, 32gb emmc are known to break at 20TB-30TB.

**Comment by nhan000** [comment]
I was able to upload a 18.1 GB, 360-degree video.

**Comment by stuzor** [comment]
Was there ever a solution here? Are we able to mount the google photos app directory to the external drive as well? I am looking to do a massive upload, around 4TB and would like that to not be written to internal storage.

**Comment by milchalpenl** [comment]
I don't think NAND is being used as temp before uploading but perhaps used for storing some kind of SHA256 hash/index etc.

I must admit that I have not done complete analysis like vegedb. But my experience is this

Scenario

- I have  3 accounts in  pixel
- Connected to USB HDD (500GB) with power etc
- Once connected etc. it is in preparing status (USB-HDD status blinking i.e reading the disk)
- Uploads start after 1 hour etc
- Situation: Lets say it is finished uploading all files from 500GB to account 1
- Now. I change backup to account 2 
- Uploads start immediately. Delay is only in a few minutes.


WILD GUESS

- Maybe it downloads hashes from account1 (quick process) to compare it to local files hashes (time consuming)
- For the second account, since localfilehash has generation is complete, account 2 (download online hashes - quick) - then start upload immediately.

**Comment by divinity76** [comment]
On rooted phones, I think we can hook `com.google.android.apps.photos` 's fopen()/open() to redirect `/data/data/com.google.android.apps.photos/databases/gphotos0.db` to wherever we want it?

I have done this on Microsoft Windows in the past, worked like a charm, and in theory it should work on Android too.. 

Fwiw I don't have a Pixel 1, but bought one on Ebay, tracking says
>Arrives Mon, Apr 6 - Mon, Apr 13

which is about 25 days from now..

**Comment by divinity76** [comment]
Btw have you guys already tested symlinking schenanigans? make `/data/data/com.google.android.apps.photos/databases/gphotos0.db`  a symlink to a file on your external drive?

**Comment by divinity76** [comment]
or maybe bind-mount /data/data/com.google.android.apps.photos/databases to the external drive?

**Comment by master-hax** [comment]
>or maybe bind-mount /data/data/com.google.android.apps.photos/databases to the external drive?

yes that's what i would recommend - though probably better to mount the entire `/data/data/com.google.android.apps.photos` to keep the data consistent

>Fwiw I don't have a Pixel 1, but bought one on Ebay

be careful, there are fake pixels being sold on ebay - not sure exactly how they are manufactured but they have mismatched IMEI & serial numbers.

check the device by visiting https://store.google.com/us/device-authenticity?id=xxxxxxxxxxxxxxx&hl=en-US

replace `xxxxxxxxxxxxxxx` with the serial number first, and then with the IMEI - then make sure all the values match with the device in your hands.

i received two devices from an ebay seller with the wrong serial number/IMEI, wrong color, & wrong size.

**Comment by divinity76** [comment]
@master-hax 
> be careful, there are fake pixels being sold on ebay - 
>(...)
> i received two devices from an ebay seller with the wrong serial number/IMEI, wrong color, & wrong size.

Wow thanks for the warning. What happened with the fakes? Did you report counterfeit to Ebay? Did you get a refund? tell me everything please

**Comment by divinity76** [comment]
Edit: rereading the conversation above, it needs to be at least as big as your biggest media file 🤔 

Any idea how big this `/data/data/com.google.android.apps.photos` folder need to be? 

If its just a couple hundred megabytes, maybe we could create a ram-disk and bind-mount /data/data/com.google.android.apps.photos to the ram disk
(the phone have 4GB ram after all 🤔 )

Btw I got my Pixel 1, validated it as genuine, and rooted it 😄 Haven't actually tried getting free storage yet, but I should hurry up because

>You've used all 15 GB of your Google Account storage. Starting on 19 Apr 2026, you will be unable to send or receive emails in Gmail.

**Comment by master-hax** [comment]
i added a new script in the repo that solves this - `yoink_gphotos.sh` makes the app run off the external drive instead of the internal storage.

note that it needs to be an ext4 drive & there's no un-yoinking script at the moment, so just restart or power off the device when done.

please give it a try @vegedb

**Comment by elfman03** [comment]
This looks promising; I'll have to put it on my list to test in the next few weeks and report back on.  

Does anyone have a guess how large of a external drive you need to use for this?    Like a 16GB thumb drive would be sufficient if I'm not uploading any individual file larger than a few gigs or its going to be something likely to creep up in size as more and more stuff uploads.  I worry a bit that the app might get confused if the yoinked storage target gets full while internal storage is reporting that space is available.

It looks like there is some traffic to com.android.providers.media when the scan intent happens.  Its not big like photos but I don't know if it has a lot of small ops.

My goal is to use a thumb drive for app storage space to cut down on NAND thrash and a CIFS network share as the source of images.

**Comment by master-hax** [comment]
when I was backing up ~800GiB, my pixel read ~2.5TiB from the drive and wrote ~1GiB to the drive. so I think a 16GiB drive should be fine.
